Add support for licensed sub feature privileges

legrego opened the pull request:
## Plugin API changes

Allows `independent` sub-feature privileges to be licensed.

Features support defining a set of valid licenses for which they are available. While this works for conditionally supporting top-level features, it doesn't scale to sub-feature privileges.

Currently, there is no way to define a sub-feature privilege which is only available at a certain license level. This PR introduces a `minimumLicense` property on each sub-feature privilege, so that consumers can choose the set of valid licenses for their sub-feature privileges.

A concrete example is Reporting: there are different report types offered at different license levels. PDF reports are a Platinum feature, so an administrator configuring roles in a Gold cluster shouldn't be able to toggle the PDF report privilege.

With this PR, licensed sub-feature privileges will only be registered with Elasticsearch when the `minimumLicense` is satisfied. Further, the sub-feature privilege will only be included into the primary feature privileges when the `minimumLicense` is satisfied. The privilege registration system is already configured to listen to license changes at runtime, so the set of available/registered sub-feature privileges will always be kept in sync.

Resolves https://github.com/elastic/kibana/issues/80385

cc @tsullivan 

### Checklist

- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/)
* Commit: 6e8be27d045eaffa7073633ad2f004f519174d47
* This comment will update when the build is complete

### Failed CI Steps
* [X-Pack Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/execution/node/338/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/","number":"82282","commit":"6e8be27d045eaffa7073633ad2f004f519174d47"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/)
* Commit: 7c49be8b3f05e0c29ad84dd0ce669d666f891631

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### async chunks size
| id | [before](https://github.com/elastic/kibana/commit/decb9dd1e11a63e450da21423df894dadbe3c332) | [after](https://github.com/elastic/kibana/commit/7c49be8b3f05e0c29ad84dd0ce669d666f891631) | diff |
| --- | --- | --- | --- |
| `security` | 1.0MB | 1.0MB | +154.0B |

#### page load bundle size
| id | [before](https://github.com/elastic/kibana/commit/decb9dd1e11a63e450da21423df894dadbe3c332) | [after](https://github.com/elastic/kibana/commit/7c49be8b3f05e0c29ad84dd0ce669d666f891631) | diff |
| --- | --- | --- | --- |
| `security` | 172.0KB | 172.2KB | +132.0B |

### History
* :broken_heart: [Build #82282](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/) failed 6e8be27d045eaffa7073633ad2f004f519174d47

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/","number":"82282","commit":"6e8be27d045eaffa7073633ad2f004f519174d47"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/","number":"82416","commit":"7c49be8b3f05e0c29ad84dd0ce669d666f891631"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/)
* Commit: 134af9521a29d0361e1fd5ef244e1e6d68de605f

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### async chunks size
| id | [before](https://github.com/elastic/kibana/commit/7b1b36e38d0a21712ccb1b805ac0416c372f37b5) | [after](https://github.com/elastic/kibana/commit/134af9521a29d0361e1fd5ef244e1e6d68de605f) | diff |
| --- | --- | --- | --- |
| `security` | 1.0MB | 1.0MB | +154.0B |

#### page load bundle size
| id | [before](https://github.com/elastic/kibana/commit/7b1b36e38d0a21712ccb1b805ac0416c372f37b5) | [after](https://github.com/elastic/kibana/commit/134af9521a29d0361e1fd5ef244e1e6d68de605f) | diff |
| --- | --- | --- | --- |
| `security` | 172.0KB | 172.2KB | +132.0B |

### History
* :green_heart: [Build #82416](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/) succeeded 7c49be8b3f05e0c29ad84dd0ce669d666f891631
* :broken_heart: [Build #82282](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/) failed 6e8be27d045eaffa7073633ad2f004f519174d47

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/","number":"82282","commit":"6e8be27d045eaffa7073633ad2f004f519174d47"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/","number":"82416","commit":"7c49be8b3f05e0c29ad84dd0ce669d666f891631"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/","number":"82599","commit":"134af9521a29d0361e1fd5ef244e1e6d68de605f"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82880/)
* Commit: f915139fc56c794b760b4e426a9bd96273eb516f

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### async chunks size
| id | [before](https://github.com/elastic/kibana/commit/b0a56b940906a82b76ca7ec1cd4b2f795bb037b4) | [after](https://github.com/elastic/kibana/commit/f915139fc56c794b760b4e426a9bd96273eb516f) | diff |
| --- | --- | --- | --- |
| `security` | 1.0MB | 1.0MB | +154.0B |

#### page load bundle size
| id | [before](https://github.com/elastic/kibana/commit/b0a56b940906a82b76ca7ec1cd4b2f795bb037b4) | [after](https://github.com/elastic/kibana/commit/f915139fc56c794b760b4e426a9bd96273eb516f) | diff |
| --- | --- | --- | --- |
| `security` | 172.0KB | 172.2KB | +132.0B |

### History
* :green_heart: [Build #82599](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/) succeeded 134af9521a29d0361e1fd5ef244e1e6d68de605f
* :green_heart: [Build #82416](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/) succeeded 7c49be8b3f05e0c29ad84dd0ce669d666f891631
* :broken_heart: [Build #82282](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/) failed 6e8be27d045eaffa7073633ad2f004f519174d47

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/","number":"82282","commit":"6e8be27d045eaffa7073633ad2f004f519174d47"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/","number":"82416","commit":"7c49be8b3f05e0c29ad84dd0ce669d666f891631"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/","number":"82599","commit":"134af9521a29d0361e1fd5ef244e1e6d68de605f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82880/","number":"82880","commit":"f915139fc56c794b760b4e426a9bd96273eb516f"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/kibana-security (Team:Security)

azasypkin commented on x-pack/plugins/security/common/licensing/license_service.ts:
```diff
@@ -36,6 +37,8 @@ export class SecurityLicenseService {
 
         isEnabled: () => this.isSecurityEnabledFromRawLicense(rawLicense),
 
+        getType: () => rawLicense?.type,
```
I think from the technical side it's fine to expose this now as Licensing plugin already became an integral part of the core Kibana functionality. I'm totally fine with that approach.

The only purely theoretical thing that I have mixed feelings about is that there will no longer be a single place where we can see what "security features" particular license enables. It's something we can think about later though, it shouldn't slow us down in any way.

legrego replied:
> The only purely theoretical thing that I have mixed feelings about is that there will no longer be a single place where we can see what "security features" particular license enables. It's something we can think about later though, it shouldn't slow us down in any way.

That's a fair point. Although I would argue that these aren't "security features", but rather licensed features that are securable.

azasypkin replied:
> Although I would argue that these aren't "security features", but rather licensed features that are securable.

++, sorry used a poor wording, I meant more like "features gated by security".

pgayvallet commented on x-pack/plugins/features/common/sub_feature.ts:
```diff
@@ -68,6 +69,12 @@ export interface SubFeaturePrivilegeConfig
    */
   includeIn: 'all' | 'read' | 'none';
+
+  /**
+   * The minimum supported license level for this sub-feature privilege.
+   * If no license level is supplied, then this privilege will be available for all valid licenses.
+   */
+  minimumLicense?: LicenseType;
```
`If no license level is supplied, then this privilege will be available for all valid licenses`

I would expect the base feature's `minimumLicense`  to be applied instead?

legrego replied:
Good point, I'll update the comment to make this clearer

pgayvallet commented on x-pack/plugins/features/server/feature_registry.test.ts:
```diff
@@ -1280,6 +1280,123 @@ describe('FeatureRegistry', () => {
+      const featureRegistry = new FeatureRegistry();
+      expect(() => {
+        featureRegistry.registerKibanaFeature(feature1);
+      }).toThrowErrorMatchingInlineSnapshot(
+        `"child \\"subFeatures\\" fails because [\\"subFeatures\\" at position 0 fails because [child \\"privilegeGroups\\" fails because [\\"privilegeGroups\\" at position 0 fails because [child \\"privileges\\" fails because [\\"privileges\\" at position 0 fails because [child \\"minimumLicense\\" fails because [\\"minimumLicense\\" is not allowed]]]]]]]"`
+      );
```
I love our validation errors. So readable, and it helps so much to understand what failed to validate.

pgayvallet commented on x-pack/plugins/features/server/routes/index.ts:
```diff
@@ -28,20 +28,37 @@ export function defineRoutes({ router, featureRegistry }: RouteDefinitionParams)
           )
-          .map((feature) => feature.toRaw()),
+          .map((feature) => {
+            const raw = feature.toRaw();
+            if (!currentLicense || request.query.ignoreValidLicenses) {
+              return raw;
+            }
+
+            raw.subFeatures?.forEach((subFeature) => {
+              subFeature.privilegeGroups.forEach((group) => {
+                group.privileges = group.privileges.filter(
+                  (privilege) =>
+                    !privilege.minimumLicense || currentLicense.hasAtLeast(privilege.minimumLicense)
+                );
+              });
+            });
```
NIT: I know we kinda already were doing it, but I feel that this kind of logic would be better at a lower level than directly in the route handler?

`featureRegistry.getAllKibanaFeatures(currentLicense, ignoreValidLicenses)` maybe?

legrego replied:
I had that same thought, but then got lazy. Happy to explore this option though, it shouldn't be that much work.

jportner replied:
+1 to Pierre's comment.

pgayvallet commented on x-pack/plugins/security/server/authorization/privileges/feature_privilege_iterator/sub_feature_privilege_iterator.ts:
```diff
@@ -4,14 +4,19 @@
+  feature: KibanaFeature,
+  licenseType: LicenseType
 ): IterableIterator<SubFeaturePrivilegeConfig> {
   for (const subFeature of feature.subFeatures) {
     for (const group of subFeature.privilegeGroups) {
-      yield* group.privileges;
+      yield* group.privileges.filter(
+        (privilege) => !privilege.minimumLicense || privilege.minimumLicense <= licenseType
+      );
```
TIL we are still using generators 😅

legrego replied:
Should we not be using generators anymore?

azasypkin replied:
Yeah, I'm curious as well. We use them in quite a few places in security and I love them! :stuck_out_tongue:

pgayvallet replied:
Generators are fine. Using them for synchronous logic where `map/reduce/filter/flatMap` would be able to generate an array as (more?) easily seems overkill to me, but that's only personal preference, don't mind this comment :)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83725/)
* Commit: 5fbe2ebe379057523aace73ecce0a37841d74794
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83725/execution/node/389/log)

### History
* :green_heart: [Build #82880](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82880/) succeeded f915139fc56c794b760b4e426a9bd96273eb516f
* :green_heart: [Build #82599](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/) succeeded 134af9521a29d0361e1fd5ef244e1e6d68de605f
* :green_heart: [Build #82416](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/) succeeded 7c49be8b3f05e0c29ad84dd0ce669d666f891631
* :broken_heart: [Build #82282](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/) failed 6e8be27d045eaffa7073633ad2f004f519174d47

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/","number":"82282","commit":"6e8be27d045eaffa7073633ad2f004f519174d47"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/","number":"82416","commit":"7c49be8b3f05e0c29ad84dd0ce669d666f891631"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/","number":"82599","commit":"134af9521a29d0361e1fd5ef244e1e6d68de605f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82880/","number":"82880","commit":"f915139fc56c794b760b4e426a9bd96273eb516f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83725/","number":"83725","commit":"5fbe2ebe379057523aace73ecce0a37841d74794"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83728/)
* Commit: 502ad7fcea5c9c2d7ffe6b70b0520463fbf025aa

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### async chunks size
| id | [before](https://github.com/elastic/kibana/commit/7e34bf2dbcbce4fee9e3508270f78dcd43e09fca) | [after](https://github.com/elastic/kibana/commit/502ad7fcea5c9c2d7ffe6b70b0520463fbf025aa) | diff |
| --- | --- | --- | --- |
| `security` | 1.0MB | 1.0MB | +154.0B |

#### page load bundle size
| id | [before](https://github.com/elastic/kibana/commit/7e34bf2dbcbce4fee9e3508270f78dcd43e09fca) | [after](https://github.com/elastic/kibana/commit/502ad7fcea5c9c2d7ffe6b70b0520463fbf025aa) | diff |
| --- | --- | --- | --- |
| `security` | 172.0KB | 172.2KB | +132.0B |

### History
* :broken_heart: [Build #83725](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83725/) failed 5fbe2ebe379057523aace73ecce0a37841d74794
* :green_heart: [Build #82880](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82880/) succeeded f915139fc56c794b760b4e426a9bd96273eb516f
* :green_heart: [Build #82599](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/) succeeded 134af9521a29d0361e1fd5ef244e1e6d68de605f
* :green_heart: [Build #82416](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/) succeeded 7c49be8b3f05e0c29ad84dd0ce669d666f891631
* :broken_heart: [Build #82282](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/) failed 6e8be27d045eaffa7073633ad2f004f519174d47

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82282/","number":"82282","commit":"6e8be27d045eaffa7073633ad2f004f519174d47"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/","number":"82416","commit":"7c49be8b3f05e0c29ad84dd0ce669d666f891631"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/","number":"82599","commit":"134af9521a29d0361e1fd5ef244e1e6d68de605f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82880/","number":"82880","commit":"f915139fc56c794b760b4e426a9bd96273eb516f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83725/","number":"83725","commit":"5fbe2ebe379057523aace73ecce0a37841d74794"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83728/","number":"83728","commit":"502ad7fcea5c9c2d7ffe6b70b0520463fbf025aa"}]}
PIPELINE-->

jportner commented on x-pack/plugins/features/server/feature_schema.ts:
```diff
@@ -53,10 +55,11 @@ const kibanaPrivilegeSchema = Joi.object({
   id: Joi.string().regex(subFeaturePrivilegePartRegex).required(),
   name: Joi.string().required(),
   includeIn: Joi.string().allow('all', 'read', 'none').required(),
+  minimumLicense: Joi.string().valid(...validLicenses),
```
Observation: We don't attempt to prevent consumers from registering sub-feature privileges with a lower minimum license level than their parent privileges.

I don't think that we necessarily need to add that level of validation, but I wanted to point it out for posterity / in case it was an oversight.

jportner replied:
Another observation: The entire sub-feature privileges feature (ha!) is only available in Gold+. So, really, 'basic', 'standard', and 'gold' shouldn't be valid options... WDYT?

legrego replied:
> We don't attempt to prevent consumers from registering sub-feature privileges with a lower minimum license level than their parent privileges.
> 
> I don't think that we necessarily need to add that level of validation, but I wanted to point it out for posterity / in case it was an oversight.

I agree, I feel like validation for this is overkill, but thanks for pointing this out!

> The entire sub-feature privileges feature (ha!) is only available in Gold+. So, really, 'basic', 'standard', and 'gold' shouldn't be valid options... WDYT?

Makes a lot of sense, I'll make this change!

jportner commented on x-pack/plugins/features/server/feature_registry.test.ts:
```diff
@@ -1305,6 +1423,97 @@ describe('FeatureRegistry', () => {
+      const registry = new FeatureRegistry();
+      features.forEach((f) => registry.registerKibanaFeature(f));
+
+      it('returns all features by default', () => {
+        const result = registry.getAllKibanaFeatures();
+        expect(result).toHaveLength(3);
+      });
```
Nit
```suggestion
      it('returns all features and sub-feature privileges by default', () => {
        const result = registry.getAllKibanaFeatures();
        expect(result).toHaveLength(3);

        const [, , withSubFeature] = result;
        expect(withSubFeature.subFeatures).toHaveLength(1);
        expect(withSubFeature.subFeatures[0].privilegeGroups).toHaveLength(1);
        expect(withSubFeature.subFeatures[0].privilegeGroups[0].privileges).toHaveLength(1);
      });
```

jportner commented on x-pack/plugins/features/server/feature_registry.test.ts:
```diff
@@ -1305,6 +1423,97 @@ describe('FeatureRegistry', () => {
+        expect(ids).toEqual(['gold-feature', 'unlicensed-feature', 'with-sub-feature']);
+
+        const [, , withSubFeature] = result;
+        expect(withSubFeature.subFeatures).toMatchInlineSnapshot(`
+          Array [
+            SubFeature {
+              "config": Object {
+                "name": "licensed-sub-feature",
+                "privilegeGroups": Array [
+                  Object {
+                    "groupType": "independent",
+                    "privileges": Array [],
+                  },
+                ],
+              },
+            },
+          ]
+        `);
```
Optional nit to get rid of inline snapshot 🙂 
```suggestion
        expect(withSubFeature.subFeatures).toHaveLength(1);
        expect(withSubFeature.subFeatures[0].privilegeGroups).toHaveLength(1);
        expect(withSubFeature.subFeatures[0].privilegeGroups[0].privileges).toHaveLength(0);
```

jportner commented on x-pack/plugins/features/server/oss_features.test.ts:
```diff
@@ -50,6 +50,7 @@ Array [
       const privileges = [];
       for (const featurePrivilege of featurePrivilegeIterator(new KibanaFeature(featureConfig), {
         augmentWithSubFeaturePrivileges: true,
+        licenseType: 'enterprise',
```
Warning: this comment may actually be complete and utter nonsense.

Changing this to `licenseType: 'basic'`, the tests still pass. It looks like this is expected as we don't have any OSS features with minimum-licensed sub-feature privileges. However, this may change in the future. Perhaps we should add another set of test cases with the basic license to be safe?

legrego replied:
You're right that we don't have any OSS features with licensed minimum sub-feature privileges yet. We will though if/when reporting takes advantage of this, though.

The addition of `licenseType` in this test was solely to satisfy the new function signature.

> Perhaps we should add another set of test cases with the basic license to be safe?

Yeah I think that's worthwhile

jportner commented on x-pack/plugins/security/server/authorization/privileges/feature_privilege_iterator/feature_privilege_iterator.test.ts:
```diff
@@ -818,6 +826,133 @@ describe('featurePrivilegeIterator', () => {
     ]);
   });
 
+  it('excludes sub feature privileges when the minimum license is not met', () => {
```
Optional nit: This test suite is rather large already, and this test case is exercising functionality that's encapsulated by `subFeaturePrivilegeIterator`. It may be better to split this test case out into a `sub_feature_privilege_iterator.test.tsx` instead.

legrego replied:
Yeah, I know you're right 😬

legrego replied:
I've opened https://github.com/elastic/kibana/issues/83457 to address this in a followup. I'm trying to reduce the number of tasks I have in flight, and would rather not hold this work up on that effort. I do think it's worthwhile though!

jportner commented on x-pack/plugins/security/server/authorization/privileges/privileges.test.ts:
```diff
@@ -1705,4 +1722,502 @@ describe('subFeatures', () => {
     });
   });
+
+  describe(`when license allows subfeatures, but not a specific sub feature`, () => {
```
Optional nit: Same as comment above.

jportner reviewed (commented):
A few nits below, but overall looking great!

One thing I'm musing about: I wonder what would happen if, for instance, you:
1. Have a cluster with a Platinum license
2. Create a role using a Platinum sub-feature privilege
3. Downgrade to a Gold license

Would the role continue working as specified, and you just wouldn't be able to change it? If not, what would happen? I couldn't easily test this locally as we don't have any consumers of this enhancement yet 🙂

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/88013/)
* Commit: 9fac73336e84eeb18191df1b6ecc87bf17bae3d9

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/3ba7758a4fad65d0df62eafb4ffd03084195eed0) | [after](https://github.com/elastic/kibana/commit/9fac73336e84eeb18191df1b6ecc87bf17bae3d9) | diff |
| --- | --- | --- | --- |
| `security` | 784.3KB | 784.5KB | +137.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/3ba7758a4fad65d0df62eafb4ffd03084195eed0) | [after](https://github.com/elastic/kibana/commit/9fac73336e84eeb18191df1b6ecc87bf17bae3d9) | diff |
| --- | --- | --- | --- |
| `security` | 162.4KB | 162.5KB | +118.0B |

### History
* :green_heart: [Build #83728](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83728/) succeeded 502ad7fcea5c9c2d7ffe6b70b0520463fbf025aa
* :broken_heart: [Build #83725](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83725/) failed 5fbe2ebe379057523aace73ecce0a37841d74794
* :green_heart: [Build #82880](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82880/) succeeded f915139fc56c794b760b4e426a9bd96273eb516f
* :green_heart: [Build #82599](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/) succeeded 134af9521a29d0361e1fd5ef244e1e6d68de605f
* :green_heart: [Build #82416](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/) succeeded 7c49be8b3f05e0c29ad84dd0ce669d666f891631

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82416/","number":"82416","commit":"7c49be8b3f05e0c29ad84dd0ce669d666f891631"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/","number":"82599","commit":"134af9521a29d0361e1fd5ef244e1e6d68de605f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82880/","number":"82880","commit":"f915139fc56c794b760b4e426a9bd96273eb516f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83725/","number":"83725","commit":"5fbe2ebe379057523aace73ecce0a37841d74794"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83728/","number":"83728","commit":"502ad7fcea5c9c2d7ffe6b70b0520463fbf025aa"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/88013/","number":"88013","commit":"9fac73336e84eeb18191df1b6ecc87bf17bae3d9"}]}
PIPELINE-->

legrego commented:
@jportner I think I've addressed your feedback with the exception of https://github.com/elastic/kibana/pull/80905#discussion_r513009129, which I can do in a followup. Ready for another round of reviews when you get a chance, but no rush.

jportner reviewed (approved):
LGTM!

If you decided not to do this / put it off for another time, please ignore:
Just wanted to point out that there aren't any changes regarding https://github.com/elastic/kibana/pull/80905#discussion_r510217653, in case that was an oversight!

legrego commented:
> If you decided not to do this / put it off for another time, please ignore:
> Just wanted to point out that there aren't any changes regarding [#80905 (comment)](https://github.com/elastic/kibana/pull/80905#discussion_r510217653), in case that was an oversight!

Oh good catch! I had done all of the work to move this logic, but never removed the old. I'll take care of that now. Thanks 👍

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/88092/)
* Commit: 8a91f2ba1a510ec6c4f5e9df33cdc86a72ede05e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/3ba7758a4fad65d0df62eafb4ffd03084195eed0) | [after](https://github.com/elastic/kibana/commit/8a91f2ba1a510ec6c4f5e9df33cdc86a72ede05e) | diff |
| --- | --- | --- | --- |
| `security` | 784.3KB | 784.5KB | +137.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/3ba7758a4fad65d0df62eafb4ffd03084195eed0) | [after](https://github.com/elastic/kibana/commit/8a91f2ba1a510ec6c4f5e9df33cdc86a72ede05e) | diff |
| --- | --- | --- | --- |
| `security` | 162.4KB | 162.5KB | +118.0B |

### History
* :green_heart: [Build #88013](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/88013/) succeeded 9fac73336e84eeb18191df1b6ecc87bf17bae3d9
* :green_heart: [Build #83728](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83728/) succeeded 502ad7fcea5c9c2d7ffe6b70b0520463fbf025aa
* :broken_heart: [Build #83725](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83725/) failed 5fbe2ebe379057523aace73ecce0a37841d74794
* :green_heart: [Build #82880](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82880/) succeeded f915139fc56c794b760b4e426a9bd96273eb516f
* :green_heart: [Build #82599](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/) succeeded 134af9521a29d0361e1fd5ef244e1e6d68de605f

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82599/","number":"82599","commit":"134af9521a29d0361e1fd5ef244e1e6d68de605f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/82880/","number":"82880","commit":"f915139fc56c794b760b4e426a9bd96273eb516f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83725/","number":"83725","commit":"5fbe2ebe379057523aace73ecce0a37841d74794"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/83728/","number":"83728","commit":"502ad7fcea5c9c2d7ffe6b70b0520463fbf025aa"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/88013/","number":"88013","commit":"9fac73336e84eeb18191df1b6ecc87bf17bae3d9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/88092/","number":"88092","commit":"8a91f2ba1a510ec6c4f5e9df33cdc86a72ede05e"}]}
PIPELINE-->
